### ISSUE TITLE
[feature] Use environment variables inside the HTML

### ISSUE THREAD

**Issue by GaelS** [parcel-bundler/parcel#1209]
🙋 feature request!

First thing first: Thanks for the nice work ! :+1: 

I have a use case where an anchor tag redirects to an external URL based on the environment (`localhost:3333` if in *DEV* for instance and `https://whatever.com` in *PROD*). 

Therefore, it would be really convenient to be able to use the environment variables directly inside the `index.html` without going through any javascript code.

Just like this : `<a href=%MY_ENV_VARIABLE%>Ok</a>`

The `%` are used randomly just to show some kind of formatting to handle ENV variables at build time. 

Maybe I have missed something but I was not able to find any kind of documentation or issues (open or close) dealing with this kind of things. Maybe I'm totally missing something but with this type of use cases, this feature would be very convenient.

**Comment by goyney** [comment]
+1

We would like to use different favicons to signify our dev environment versus our production environment so that when we have many tabs open, it's easy to pick out where you're looking. Having access to an env variable in the HTML would allow us to do this easily.

**Comment by adamelliotfields** [comment]
Would also be useful for additional script tags (service worker, Google analytics, etc) that you don't need when developing/designing your app; but obviously need when deployed.

I _think_ this could be resolved by allowing the entry point to be a template file like Handlebars, EJS, or Pug. Parcel automatically sets the `NODE_ENV` to `production` which could be passed to the template at build time.

Those of us coming from Webpack got used to `html-webpack-plugin`, so it would be great to see Parcel incorporate similar functionality, while still maintaining the exceptional developer experience it offers over Webpack.

**Comment by breier** [comment]
You can use `<%= MY_ENV_VAR %>`
`<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <link rel="icon" href="<%= BASE_URL %>favicon.ico">
    <title><%= VUE_APP_TITLE %></title>
</head>`

**Comment by mrcoles** [comment]
I was able to get environment variables within my HTML using PostHTML with the following steps:

1. Set some environment variables, e.g., for dev, I created `.env.development` in the root of my project as such:

    ```
    HOME_PATH=/index.html
    ```

2. Add the posthtml-expressions library:

    ```bash
    yarn add -D posthtml-expressions
    ```

3. Add a posthtml config file—important that it’s a JS file and not just JSON—posthtml.config.js (the reason for using these as JS files is not  really documented currently, but the ability to access process.env here is why I’m using it):

    ```javascript
    module.exports = {
      plugins: {
        "posthtml-expressions": {
          locals: {
            HOME_URL: process.env.HOME_URL
          }
        }
      }
    };
    ```

4. Update my HTML to use the [posthtml-expressions syntax](https://github.com/posthtml/posthtml-expressions):

    ```html
    <a href="{{ HOME_PATH || '/' }}">home</a>
    ```

And voila, you have environment variables reflected in your HTML.

---

Also, @breier in your example, are you using some extra asset-type handler to handle that ruby syntax `<%= … %>`?

**Comment by breier** [comment]
@mrcoles I'm used `vue-cli` to start a project using `webpack` template. No additions.

**Comment by devongovett** [comment]
This isn't something we'll support out of the box since it isn't standard HTML syntax. But you could easily use a PostHTML plugin for this, or another template language like Pug.

**Comment by vicpon** [comment]
> @mrcoles I've used `vue-cli` to start a project using `webpack` template. No additions.

Are you using the vue cli to have process.env available? If so, how? Thanks

**Comment by dandv** [comment]
@mrcoles: is it HOME_URL or HOME_PATH?

**Comment by mrcoles** [comment]
> is it HOME_URL or HOME_PATH?

@dandv, sorry that was a typo, I updated my comment so all `HOME_PATH` references are `HOME_URL`. Of course, feel free to use any variable names you want in your code :shipit:

**Comment by breier** [comment]
> > @mrcoles I've used `vue-cli` to start a project using `webpack` template. No additions.
> 
> @breier Are you using the vue cli to have process.env available? If so, how? Thanks

Sorry for the really late reply, but for the record, here it is...
https://cli.vuejs.org/guide/mode-and-env.html#environment-variables
>Note that only NODE_ENV, BASE_URL, and variables that start with VUE_APP_ will be statically embedded into the client bundle with webpack.DefinePlugin. It is to avoid accidentally exposing a private key on the machine that could have the same name.
Basically, any custom env var just have to start with VUE_APP_ in order to be auto loaded and assimilated.

I hope it helps someone ;)

**Comment by zanona** [comment]
@mrcoles, were you able to use `NODE_ENV` though?
It seems to be set to `production` regardless of what's used.

#### posthtml.config.js
```js
console.log(process.env.NODE_ENV)
module.expors = {};
```

```sh
NODE_ENV=test npx parcel build --no-cache index.html
console: production
⠸ Building index.html...
```




> If you’re using something that runs JS (like react) as mentioned you can just reference `process.env` in your code. However, I was able to get environment variables within my regular HTML files using PostHTML with the following steps: 
>     3. Add a posthtml config file—important that it’s a JS file and not just JSON—`posthtml.config.js` (the reason for using these as JS files is not  really documented currently, but the ability to access process.env here is why I’m using it):
>        ```js
>        module.exports = {
>          plugins: {
>            "posthtml-expressions": {
>              locals: {
>                HOME_URL: process.env.HOME_URL
>              }
>            }
>          }
>        };
>        ```

**Comment by talor-hammond** [comment]
> If you’re using something that runs JS (like react) as mentioned you can just reference `process.env` in your code. However, I was able to get environment variables within my regular HTML files using PostHTML with the following steps:
> 
> 1. Set some [environment variables](https://parceljs.org/env.html), e.g., I created `.env.development` in the root of my project as such:
>    ```
>    HOME_URL=/index.html
>    ```
> 2. Add the posthtml-expressions library:
>    ```shell
>    yarn add -D posthtml-expressions
>    ```
> 3. Add a posthtml config file—important that it’s a JS file and not just JSON—`posthtml.config.js` (the reason for using these as JS files is not  really documented currently, but the ability to access process.env here is why I’m using it):
>    ```js
>    module.exports = {
>      plugins: {
>        "posthtml-expressions": {
>          locals: {
>            HOME_URL: process.env.HOME_URL
>          }
>        }
>      }
>    };
>    ```
> 4. Update my HTML to use the [posthtml-expressions syntax](https://github.com/posthtml/posthtml-expressions):
>    ```
>    <a href="{{ HOME_URL || '/' }}">home</a>
>    ```
> 5. In order to try it out in dev `yarn start` and to build it for production `yarn build` taking into account different environments, I have the following in my `package.json`:
>    ```json
>    {
>      "scripts": {
>        "start": "yarn parcel src/*.html",
>        "build": "NODE_ENV=prod parcel build --public-url ./ src/index.html",
>      }
>    }
>    ```
> 
> And voila, you have environment variables reflected in your HTML.
> 
> _Also, one gotcha: I’m not seeing changes to .env triggering a re-build. You’ll have to edit something to trigger a re-build or stop parcel, `rm -r .cache`, and then start parcel._
> 
> @breier in your example, are you using some extra asset-type handler to handle that ruby syntax `<%= … %>`?

Doesn't seem to work with `parcel` v2 unfortunately ):
Similar setup -- node envs aren't being recognised in `posthtml.config.js`

**Comment by talor-hammond** [comment]
> If you’re using something that runs JS (like react) as mentioned you can just reference `process.env` in your code. However, I was able to get environment variables within my regular HTML files using PostHTML with the following steps:
> 
> 1. Set some [environment variables](https://parceljs.org/env.html), e.g., I created `.env.development` in the root of my project as such:
>    ```
>    HOME_URL=/index.html
>    ```
> 2. Add the posthtml-expressions library:
>    ```shell
>    yarn add -D posthtml-expressions
>    ```
> 3. Add a posthtml config file—important that it’s a JS file and not just JSON—`posthtml.config.js` (the reason for using these as JS files is not  really documented currently, but the ability to access process.env here is why I’m using it):
>    ```js
>    module.exports = {
>      plugins: {
>        "posthtml-expressions": {
>          locals: {
>            HOME_URL: process.env.HOME_URL
>          }
>        }
>      }
>    };
>    ```
> 4. Update my HTML to use the [posthtml-expressions syntax](https://github.com/posthtml/posthtml-expressions):
>    ```
>    <a href="{{ HOME_URL || '/' }}">home</a>
>    ```
> 5. In order to try it out in dev `yarn start` and to build it for production `yarn build` taking into account different environments, I have the following in my `package.json`:
>    ```json
>    {
>      "scripts": {
>        "start": "yarn parcel src/*.html",
>        "build": "NODE_ENV=prod parcel build --public-url ./ src/index.html",
>      }
>    }
>    ```
> 
> And voila, you have environment variables reflected in your HTML.

For Parcel v2, you will have to call `dotenv.config()` at the top of your posthml config:

```js
const dotenv = require("dotenv");

dotenv.config();

module.exports = {
  plugins: {
    "posthtml-expressions": {
      locals: {
        HOME_URL: process.env.HOME_URL,
        ...
```

**Comment by northen-mindset** [comment]
> For Parcel v2, you will have to call `dotenv.config()` at the top of your posthml config:

Unfortunately, it seems to read only `.env` file:

```
error: Error: ENOENT: no such file or directory, open 'C:\Users\User\source\repos\project\.env'
```

Whereas Parcel explicitly loads `.env.${NODE_ENV}.local` and `.env.${NODE_ENV}`:

https://github.com/parcel-bundler/parcel/blob/4df85e34d919f75cfa0d8ad79e9cc7d4fd19f716/packages/core/core/src/loadDotEnv.js#L18-L26

All this sounds like a show-stopper as of now. Switching to PUG is a bit… overkill to just inject 2-3 environment variables into HTML. I'd better consider writing a simple parcel transformer for this… or a PostHTML plugin.

**Comment by supersetapp** [comment]
I’m trying out Parcel2 and using my prior posthtml-expressions example and also building off the two prior comments from @talor-hammond and @anantakrishna where it’s identified that we need to load the environment variables ourselves, I was able to get things working by doing the following:

```javascript
const dotenv = require('dotenv');

const NODE_ENV = process.env.NODE_ENV;

const dotenvFiles = [
  '.env',
  // Don't include `.env.local` for `test` environment
  // since normally you expect tests to produce the same
  // results for everyone
  NODE_ENV === 'test' ? null : '.env.local',
  `.env.${NODE_ENV}`,
  `.env.${NODE_ENV}.local`,
].filter(Boolean);

const env = {};

for (let dotenvFile of dotenvFiles) {
  const config = dotenv.config({ path: dotenvFile });
  if (config.parsed) {
    Object.assign(env, config.parsed);
  }
}

module.exports = {
  plugins: {
    'posthtml-expressions': {
      locals: {
        HOME_URL: env.HOME_URL,
      },
    },
  },
};
```

Given this extra, brittle-feeling code and also the big disclaimers that keep popping up about how a `posthtml.config.js` is terrible for caching, is there a preferred alternate way to inject environment variables into HTML? Maybe using a different template language like pug—or just moving any of this code into a JS file? (cc @devongovett)

**Comment by KaKi87** [comment]
Hello, accessing `process.env.NODE_ENV` doesn't work for me as of Parcel 2.3.1, using either `html` or `pug`. Thanks

**Comment by alshdavid** [comment]
Seconded ^

**Comment by michaellenaghan** [comment]
Putting `process.env` vars in HTML files has been, by far, the most painful part of switching from Webpack to ParcelJS.

(I have API keys for things like Google Maps, and I want to use different keys for development and production. That means simple text substitution won't work; I need `dotenv` support.)

Here's what I'm currently doing...

First, I installed `@plasmohq/parcel-transformer-inject-env`:

```sh
npm install -D @plasmohq/parcel-transformer-inject-env
```

Then I added it to `.parcelrc`:

```json
{
  "extends": "@parcel/config-default",
  "transformers": {
    "*.html": [
      "@plasmohq/parcel-transformer-inject-env",
      "..."
    ]
  }
}
```

Then I created `.env.development` and `.env.production` files that contained the keys I needed:

```sh
GOOGLE_MAPS_API_KEY=...
```

Then I referenced those keys using `$` notation in my HTML files:

```
...?key=$GOOGLE_MAPS_API_KEY...
```

The way the transformer works may not quite suit your needs — in particular, there's no escape mechanism — but the implementation is actually quite simple; see [here](https://github.com/PlasmoHQ/plasmo/blob/0213d711d82fc595190a689e7a260de3f874d0bf/packages/parcel-transformer-inject-env/src/index.ts) and [here](https://github.com/PlasmoHQ/plasmo-utils/blob/224b3d91fcba12f5b5d78e5d73d99f7f0f0bba2e/env.ts). If you need to, you can probably create your own version pretty easily.

The advantage of this approach over `posthtml-expressions` is that you aren't using a `.js` config file, so you don't loose caching.

**Comment by henriquenovais** [comment]
Just to put the experience of someone who tried the most liked strategies of implementation in this topic:
I am using the latest parcel version as of right now, nothing else.
- I tried the pug solution to build into an html file: could not make it work out of the box (tried for about two hours)
- I used the posthtml solution but it reduced greatly the build perfomance (in my opinion) if you think of scalating your project

Best solution to me was the one that came from @michaellenaghan.

It is indeed a pain to push an environment variable through dotenv to index.html using parcel, tons of trial and error.

I applied @michaellenaghan solution here: https://github.com/henriquenovais/Typescript101.
